doc/discovery/consul_kv.md

```diff
@@ -0,0 +1,147 @@
+
+## Configuration for discovery client
+
+### Configuration for Consul KV
+
+Add following configuration in `conf/config.yaml` :
+
+```yaml
+discovery:
+  consul_kv:
+    servers:
+      - "http://127.0.0.1:8500"
+      - "http://127.0.0.1:8600"
+    prefix: "upstreams"
+    skip_keys:                    # if you need to skip special keys
+      - "upstreams/unused_api/"
+    timeout:
+      connect: 1000               # default 2000 ms
+      read: 1000                  # default 2000 ms
+      wait: 60                    # default 60 sec
+    weight: 1                     # default 1
+    fetch_interval: 5             # default 3 sec, only take effect for keepalive: false way
+    keepalive: true               # default true, use the long pull way to query consul servers
+    default_server:               # you can define default server when missing hit
+      host: "127.0.0.1"
+      port: 20999
+      metadata:
+        fail_timeout: 1           # default 1 ms
+        weight: 1                 # default 1
+        max_fails: 1              # default 1
+```
+
+And you can config it in short by default value:
+
+```yaml
+discovery:
+  consul_kv:
+    servers:
+      - "http://127.0.0.1:8500"
+```
+
+The `keepalive` has two optional values:
+
+- `true`, default and recommend value, use the long pull way to query consul servers
+- `false`, not recommend, it would use the short pull way to query consul servers, then you can set the `fetch_interval` for fetch interval
+
+### Register Http API Services
+
+Service register Key&Value template:
+
+```
+Key:    {Prefix}/{Service_Name}/{IP}:{Port}
+Value: {"weight": <Num>, "max_fails": <Num>, "fail_timeout": <Num>}
+```
+
+The register consul key use `upstreams` as prefix by default. The http api service name called `webpages` for example, and you can also use `webpages/oneteam/hello` as service name. The api instance of node's ip and port make up new key: `<IP>:<Port>`.
+
+Now, register nodes into consul:
+
+```shell
+curl \
+    -X PUT \
+    -d ' {"weight": 1, "max_fails": 2, "fail_timeout": 1}' \
+    http://127.0.0.1:8500/v1/kv/upstreams/webpages/172.19.5.12:8000
+
+curl \
+    -X PUT \
+    -d ' {"weight": 1, "max_fails": 2, "fail_timeout": 1}' \
+    http://127.0.0.1:8500/v1/kv/upstreams/webpages/172.19.5.13:8000
+```
+
+In some case, same keys exist in different consul servers.
+To avoid confusion, use the full consul key url path as service name in practice.
+
+### Upstream setting
+
+Here is an example of routing a request with a URL of "/*" to a service which named "http://127.0.0.1:8500/v1/kv/upstreams/webpages/" and use consul_kv discovery client in the registry :
+
+```shell
+$ curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
+{
+    "uri": "/*",
+    "upstream": {
+        "service_name": "http://127.0.0.1:8500/v1/kv/upstreams/webpages/",
```

Comment:
Maybe we can split service_name into address, prefix and name? To better adopt other discovery_type like etcd or nacos in the future?

Reply:
@yongboy 
What about splitting it into `discovery_center_id` and `service_name`?
Like:
```
{
"discovery_center_id": "http://127.0.0.1:8500",
"service_name": "/v1/kv/upstreams/webpages/"
}
```

Reply:
I don't think it is a good idea.

For Nginx user who used the [upsync](https://github.com/weibocom/nginx-upsync-module) module, would configrate the upstream below:

```
    upstream test {
        upsync 127.0.0.1:8500/v1/kv/upstreams/test/ upsync_timeout=6m upsync_interval=500ms upsync_type=consul strong_dependency=off;
        upsync_dump_path /usr/local/nginx/conf/servers/servers_test.conf;

        include /usr/local/nginx/conf/servers/servers_test.conf;
    }
```

You can find people need to input the fully consul key url here, eg `127.0.0.1:8500/v1/kv/upstreams/test/`.

We should make things simpler, so the service_name is enough here.

BTW, the discovery model is very important for most of us, we can add more features (eg: split service_name into another two variables) in the future PR, but not in this PR.